### PR TITLE
added tokenization tasks to serve tokenization requests

### PR DESCRIPTION
**Pull request by swith004**
Added tokenization tasks in order to serve tokenization request and return the token count

### CODE REVIEW

**Review by evaline-ju — CHANGES_REQUESTED** [review]
Thanks for the contribution! A couple asks:
- A few small unit tests would be helpful for the new module function. We should strive for tests on any new functionality. A section like for the streaming tests https://github.com/caikit/caikit-nlp/blob/main/tests/modules/text_generation/test_text_generation_tgis.py#L135-L161 on each module might be useful to reference.
- We will likely want to allow for the tokenization to also run on prompt tuned models so it'd be good to have the same update here: https://github.com/caikit/caikit-nlp/blob/main/caikit_nlp/modules/text_generation/peft_tgis_remote.py
- [After all updates] For releasing we will want this duplicated to the `release-0.3` branch in order to also release this functionality without embeddings.

File: caikit_nlp/modules/text_generation/text_generation_tgis.py [hunk 1]
```diff
@@ -298,3 +299,21 @@ def run_stream_out(
                 exponential_decay_length_penalty=exponential_decay_length_penalty,
                 stop_sequences=stop_sequences,
             )
+
+    @TokenizationTask.taskmethod()
```
[the remark this answers is not in the document]
Comment (swith004) [reply caikit_nlp/modules/text_generation/text_generation_tgis.py:303]: @evaline-ju  - can you explain about this being a "multi-task"? It seems the other task in the file are TextGenerationTask, so not clear on the changes requested here for the TokenizationTask...

[the remark this answers is not in the document]
Comment (evaline-ju) [reply caikit_nlp/modules/text_generation/text_generation_tgis.py:303]: before with only `TextGenerationTask` this was only single-task not multitask. By adding the TokenizationTask this becomes a multitask module like the link. We should make sure the module accounts for this by listing the tasks in the decorator

[the remark this answers is not in the document]
Comment (swith004) [reply caikit_nlp/modules/text_generation/text_generation_tgis.py:303]: Oh, I understand what you mean now. Thanks @evaline-ju

File: caikit_nlp/toolkit/text_generation/tgis_utils.py [hunk 2]
```diff
@@ -484,3 +485,55 @@ def stream_generate(
...
+
+        error.value_check(
+            "<NLP38899018E>",
+            len(batch_response.responses) == 1,
+            f"Got {len(batch_response.responses)} responses for a single request",
+        )
+        response = batch_response.responses[0]
+
+        return TokenizationResults(
+            token_count=response.token_count,
```
[the remark this answers is not in the document]
Comment (swith004) [reply caikit_nlp/toolkit/text_generation/tgis_utils.py:533]: we were not meaning to include the tokens here. thanks - good catch!

File: caikit_nlp/toolkit/text_generation/tgis_utils.py [hunk 3]
```diff
@@ -484,3 +485,55 @@ def stream_generate(
...
+        self,
+        text,
+    ) -> TokenizationResults:
+        """Tokenize unary input using TGIS
+
+        Args:
+            {}
+        Returns:
+            TokenizationResults
+                tokens and input token count
```
[the remark this answers is not in the document]
Comment (swith004) [reply caikit_nlp/toolkit/text_generation/tgis_utils.py]: done!

File: caikit_nlp/toolkit/text_generation/tgis_utils.py [hunk 4]
```diff
@@ -484,3 +485,55 @@ def stream_generate(
...
+        """
+
+        # In case internal client is not configured - generation
+        # cannot be done (individual modules may already check
+        # for this)
+        error.value_check(
+            "<NLP72700256E>",
```
[the remark this answers is not in the document]
Comment (swith004) [reply caikit_nlp/toolkit/text_generation/tgis_utils.py]: done!

File: caikit_nlp/toolkit/text_generation/tgis_utils.py [hunk 5]
```diff
@@ -484,3 +485,55 @@ def stream_generate(
...
+        """Tokenize unary input using TGIS
+
+        Args:
+            {}
+        Returns:
+            TokenizationResults
+                tokens and input token count
+        """
+
+        # In case internal client is not configured - generation
```
[the remark this answers is not in the document]
Comment (swith004) [reply caikit_nlp/toolkit/text_generation/tgis_utils.py]: modified this comment!

File: caikit_nlp/toolkit/text_generation/tgis_utils.py [hunk 6]
```diff
@@ -484,3 +485,55 @@ def stream_generate(
...
             )
+
+    def unary_tokenize(
+        self,
+        text,
+    ) -> TokenizationResults:
+        """Tokenize unary input using TGIS
+
+        Args:
+            {}
```
[the remark this answers is not in the document]
Comment (swith004) [reply caikit_nlp/toolkit/text_generation/tgis_utils.py]: done!

File: caikit_nlp/modules/text_generation/text_generation_tgis.py [hunk 7]
```diff
@@ -298,3 +303,21 @@ def run_stream_out(
...
+    def run_tokenizer(
+        self,
+        text: str,
+    ) -> TokenizationResults:
+        """Run tokenization task against the model running in TGIS.
+
+        Args:
+           text to tokenize
```
[the remark this answers is not in the document]
Comment (swith004) [reply caikit_nlp/modules/text_generation/text_generation_tgis.py]: thanks - done!

File: caikit_nlp/toolkit/text_generation/tgis_utils.py [hunk 8]
```diff
@@ -484,3 +485,55 @@ def stream_generate(
...
+                tokens and input token count
+        """
+
+        # In case internal client is not configured - generation
+        # cannot be done (individual modules may already check
+        # for this)
+        error.value_check(
+            "<NLP72700256E>",
+            self.tgis_client is not None,
+            "Backend must be configured and loaded for generate",
```
[the remark this answers is not in the document]
Comment (swith004) [reply caikit_nlp/toolkit/text_generation/tgis_utils.py]: thanks - edited message

File: caikit_nlp/modules/text_generation/peft_tgis_remote.py [hunk 9]
```diff
@@ -284,3 +289,21 @@ def run_stream_out(
...
+
+    @TokenizationTask.taskmethod()
+    def run_tokenizer(
+        self,
+        text: str,
+    ) -> TokenizationResults:
+        """Run tokenization task against the model running in TGIS.
+
+        Args:
+        text: str
```
[the remark this answers is not in the document]
Comment (swith004) [reply caikit_nlp/modules/text_generation/peft_tgis_remote.py]: done!

**Comment by swith004** [comment]
> one tiny nit but otherwise lgtm! You'll need a duplicate PR to `release-0.3` as well for now

@tharapalanivel should I make the duplicate PR for the `release-0.3` before or after these changes get merged?

**Comment by swith004** [comment]
> Thanks for the contribution! A couple asks:
> 
>     * A few small unit tests would be helpful for the new module function. We should strive for tests on any new functionality. A section like for the streaming tests https://github.com/caikit/caikit-nlp/blob/main/tests/modules/text_generation/test_text_generation_tgis.py#L135-L161 on each module might be useful to reference.
> 
>     * We will likely want to allow for the tokenization to also run on prompt tuned models so it'd be good to have the same update here: https://github.com/caikit/caikit-nlp/blob/main/caikit_nlp/modules/text_generation/peft_tgis_remote.py
> 
>     * [After all updates] For releasing we will want this duplicated to the `release-0.3` branch in order to also release this functionality without embeddings.

@evaline-ju - merging is still block despite Thara's approval and the changes you requested. Not sure how to proceed

**Review by evaline-ju — APPROVED** [review]

**Pull request merged**